plugin-server/src/worker/vm/upgrades/historical-export/export-historical-events-v2.ts

```diff
@@ -0,0 +1,487 @@
+): void {
+    const { methods, tasks, meta } = response
+
+    const currentPublicJobs = pluginConfig.plugin?.public_jobs || {}
+
+    if (!(INTERFACE_JOB_NAME in currentPublicJobs)) {
+        hub.promiseManager.trackPromise(hub.db.addOrUpdatePublicJob(pluginConfig.plugin_id, INTERFACE_JOB_NAME, {}))
+    }
+
+    const oldRunEveryMinute = tasks.schedule.runEveryMinute?.exec
+
+    tasks.job[INTERFACE_JOB_NAME] = {
+        name: INTERFACE_JOB_NAME,
+        type: PluginTaskType.Job,
+        // TODO: Accept timestamp as payload
+        exec: async (payload: ExportHistoricalEventsUIPayload) => {
+            // only let one export run at a time
+            const alreadyRunningExport = await getExportParameters()
+            if (!!alreadyRunningExport) {
+                return
+            }
+
+            // :TODO: Clear/invalidate old export storage somehow
+
+            const id = Math.floor(Math.random() * 10000 + 1)
+            const parallelism = Number(payload.parallelism ?? 1)
+            const boundaries = await getTimestampBoundaries(payload)
+
+            await meta.storage.set(EXPORT_RUNNING_KEY, {
+                id,
+                parallelism,
+                dateFrom: boundaries.min.toISOString(),
+                dateTo: boundaries.max.toISOString(),
+            } as ExportParams)
+
+            await coordinateHistoricExport()
+        },
+    } as unknown as PluginTask // :KLUDGE: Work around typing limitations
+
+    tasks.job['exportHistoricalEvents'] = {
+        name: 'exportHistoricalEvents',
+        type: PluginTaskType.Job,
+        exec: (payload) => exportHistoricalEvents(payload as ExportHistoricalEventsJobPayload),
+    }
+
+    tasks.schedule.runEveryMinute = {
+        name: 'runEveryMinute',
+        type: PluginTaskType.Schedule,
+        exec: async () => {
+            await oldRunEveryMinute?.()
+            await coordinateHistoricExport()
+        },
+    }
+
+    async function exportHistoricalEvents(payload: ExportHistoricalEventsJobPayload): Promise<void> {
+        if (payload.retriesPerformedSoFar >= 15) {
+            // :TODO: Should we cancel the whole export? Notify users?
+            return
+        }
+
+        // :TODO: Handle task when this export isn't active anymore.
+        // :TODO: Handle double-processing somehow?
+
+        if (payload.timestampCursor >= payload.endTime) {
+            createLog(
+                `Finished processing events between ${new Date(payload.startTime).toISOString()} and ${new Date(
+                    payload.endTime
+                ).toISOString()}`
+            )
+            await meta.storage.set(payload.statusKey, {
+                done: true,
+                progress: 1,
+            } as ExportDateStatus)
+
+            return
+        }
+
+        // :TODO: This key system is incompatible with what's done elsewhere.
+        await meta.storage.set(payload.statusKey, {
```

Comment:
Understand this is currently WIP so we don't yet have a del for this key anywhere. 

Just writing as a reminder to check we do delete all of these

Reply:
Yup - there's a TODO in there for cleanup or avoiding conflicts on the next run (as cleanup might be impossible)